src/dash-table/components/ControlledTable/index.tsx

```diff
@@ -363,69 +364,88 @@ export default class ControlledTable extends PureComponent<ControlledTableProps>
-        const selectedCols = sortNumerical(R.uniq(R.pluck(1, selected_cells)));
-
-        const minRow = selectedRows[0];
-        const minCol = selectedCols[0];
-        const maxRow = selectedRows[selectedRows.length - 1];
-        const maxCol = selectedCols[selectedCols.length - 1];
 
+        let {minRow, minCol, maxRow, maxCol} = selectionBounds(selected_cells);
         const selectingDown =
             e.keyCode === KEY_CODES.ARROW_DOWN || e.keyCode === KEY_CODES.ENTER;
         const selectingUp = e.keyCode === KEY_CODES.ARROW_UP;
         const selectingRight =
             e.keyCode === KEY_CODES.ARROW_RIGHT || e.keyCode === KEY_CODES.TAB;
         const selectingLeft = e.keyCode === KEY_CODES.ARROW_LEFT;
-
-        // If there are selections above the active cell and we are
-        // selecting down then pull down the top selection towards
-        // the active cell.
-        if (selectingDown && (active_cell as any)[0] > minRow) {
-            removeCells = selectedCols.map(col => [minRow, col]);
-        } else if (selectingDown && maxRow !== viewport.data.length - 1) {
-            // Otherwise if we are selecting down select the next row if possible.
-            targetCells = selectedCols.map(col => [maxRow + 1, col]);
-        } else if (selectingUp && (active_cell as any)[0] < maxRow) {
-            // If there are selections below the active cell and we are selecting
-            // up remove lower row.
-            removeCells = selectedCols.map(col => [maxRow, col]);
-        } else if (selectingUp && minRow > 0) {
-            // Otherwise if we are selecting up select next row if possible.
-            targetCells = selectedCols.map(col => [minRow - 1, col]);
-        } else if (selectingLeft && (active_cell as any)[1] < maxCol) {
-            // If there are selections to the right of the active cell and
-            // we are selecting left, move the right side closer to active_cell
-            removeCells = selectedRows.map(row => [row, maxCol]);
-        } else if (selectingLeft && minCol > 0) {
-            // Otherwise increase the selection left if possible
-            targetCells = selectedRows.map(row => [row, minCol - 1]);
-        } else if (selectingRight && (active_cell as any)[1] > minCol) {
-            // If there are selections to the left of the active cell and
-            // we are selecting right, move the left side closer to active_cell
-            removeCells = selectedRows.map(row => [row, minCol]);
-        } else if (selectingRight && maxCol + 1 <= columns.length - 1) {
-            // Otherwise move selection right if possible
-            targetCells = selectedRows.map(row => [row, maxCol + 1]);
+        let startRow = start_cell && start_cell.row;
+        let startCol = start_cell && start_cell.column;
+        let endRow = end_cell && end_cell.row;
+        let endCol = end_cell && end_cell.column;
+
+        if (selectingDown) {
+            if (active_cell.row > minRow) {
```

Comment:
I changed the logic here to be explicitly rectangular - ie picking out the region edges, editing these, and then recreating the `selected_cells` list from that at the end - rather than directly manipulating `selected_cells` as you had previously. I did this because it seemed both clearer and faster, I needed the new edges anyway for `end_cell`, and this way `selected_cells` will always be ordered as it appears on screen.

I considered changing these conditions from `active_cell` to `start_cell`. These are only different if you move `active_cell` after making the selection (via `tab` or `enter`), and as it stands some weird things can happen to the selection if you alter it via `shift-arrow` at that point - it's hard to predict whether `shift-down` will add a row at the bottom or remove one at the top. If I used `start_cell` instead of `active_cell` here, the cell you first clicked on would always be a corner of the selection, and `shift-arrow` would move the opposite corner. But there are two problems with that approach: (1) after you've moved `active_cell`, there's no visual indicator of which corner you started on, and (2) it's possible to shrink the selection so `active_cell` is no longer part of the selection. So in the end I left it with `active_cell`, but we could revisit that.